chaos/cluster/cluster.go

```diff
@@ -0,0 +1,219 @@
+package cluster
+
+import (
+	"bytes"
+	"fmt"
+	"math/rand"
+	"sync"
+	"time"
+
+	"github.com/golang/glog"
+	"golang.org/x/crypto/ssh"
+
+	"github.com/coreos/ktestutil/utils"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/pkg/api/v1"
+)
+
+const (
+	cmdReboot          = "sudo ifconfig eth0 down && sudo systemctl stop kubelet && sudo reboot"
+	cmdRebootWithSleep = "sudo ifconfig eth0 down && sudo systemctl stop kubelet && sleep %ds && sudo reboot"
+	cmdSystemUp        = "sudo systemctl is-active kubelet"
+)
+
+// Cluster is a simple abstraction that stores cluster nodes.
+// It allows rebooting the entire cluster / nodes.
+type Cluster struct {
+	Masters []*v1.Node
+	Workers []*v1.Node
+	// MaxDisruption defines the max no. of nodes that will be rebooted in parallel.
+	// Accepts int eg. '3' for no. of nodes, and string '30%' for percent.
+	// Defaults to 100%.
+	MaxDisruption intstr.IntOrString
+
+	client    kubernetes.Interface
+	sshClient *utils.SSHClient
+	sshConfig *utils.SSHConfig
+}
+
+// New can be called in to return a *Cluster object.
+func New(client kubernetes.Interface, opts ...Options) (*Cluster, error) {
+	cl := &Cluster{
+		client:        client,
+		sshConfig:     &utils.SSHConfig{},
+		MaxDisruption: intstr.FromString("100%"),
+	}
+	for _, opt := range opts {
+		opt(cl)
+	}
+
+	//setup ssh client
+	cl.sshClient = utils.MustNewSSHClient(cl.sshConfig)
+
+	//setup nodes
+	nodelist, err := client.CoreV1().Nodes().List(metav1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+	for i := range nodelist.Items {
+		node := &nodelist.Items[i]
+		if utils.IsMaster(node) {
+			cl.Masters = append(cl.Masters, node)
+		}
+		if utils.IsWorker(node) {
+			cl.Workers = append(cl.Workers, node)
+		}
+	}
+
+	return cl, nil
+}
+
+// RebootAll reboots all the nodes that are accessible ie. have ExternalIP.
+// rebootAfter is the time duration between stopping kubelet and issuing reboot.
+func (cl *Cluster) RebootAll(rebootAfter time.Duration) error {
```

Comment:
rebootAfter -> rebootDuration.

rebootDuration is the time for "rebooting". If you want to go into details, explain that we simulate the "reboot" duration by shutdown interface and kubelet, wait X, then do the actually reboot.